Can't build guage on Apple Silicon

I built `gauge` on MacBook Air with Apple M1 chip and it raised this

https://github.com/vincenthz/hs-gauge/blob/496a8b99e7fee3039fd89ecef7305d31eb5d5a6d/cbits/cycles.c#L55

error.

When I ran

```
❯ cabal install --lib gauge
```

and got this output:

<details>

<summary>Output</summary>

```
Resolving dependencies...
Build profile: -w ghc-8.10.7 -O1
In order, the following will be built (use -v for more details):
 - gauge-0.2.5 (lib) (requires build)
Starting     gauge-0.2.5 (lib)
Building     gauge-0.2.5 (lib)

Failed to build gauge-0.2.5.
Build log ( /Users/***/.cabal/logs/ghc-8.10.7/gg-0.2.5-a59d4712.log ):
Configuring library for gauge-0.2.5..
Preprocessing library for gauge-0.2.5..
Building library for gauge-0.2.5..
[ 1 of 43] Compiling Gauge.CSV        ( Gauge/CSV.hs, dist/build/Gauge/CSV.o, dist/build/Gauge/CSV.dyn_o )
[ 2 of 43] Compiling Gauge.ListMap    ( Gauge/ListMap.hs, dist/build/Gauge/ListMap.o, dist/build/Gauge/ListMap.dyn_o )
[ 3 of 43] Compiling Gauge.Optional   ( Gauge/Optional.hs, dist/build/Gauge/Optional.o, dist/build/Gauge/Optional.dyn_o )
[ 4 of 43] Compiling Gauge.Source.Time ( dist/build/Gauge/Source/Time.hs, dist/build/Gauge/Source/Time.o, dist/build/Gauge/Source/Time.dyn_o )
[ 5 of 43] Compiling Gauge.Time       ( Gauge/Time.hs, dist/build/Gauge/Time.o, dist/build/Gauge/Time.dyn_o )
[ 6 of 43] Compiling Gauge.Source.RUsage ( dist/build/Gauge/Source/RUsage.hs, dist/build/Gauge/Source/RUsage.o, dist/build/Gauge/Source/RUsage.dyn_o )
[ 7 of 43] Compiling Gauge.Source.GC  ( Gauge/Source/GC.hs, dist/build/Gauge/Source/GC.o, dist/build/Gauge/Source/GC.dyn_o )
[ 8 of 43] Compiling Gauge.Measurement ( Gauge/Measurement.hs, dist/build/Gauge/Measurement.o, dist/build/Gauge/Measurement.dyn_o )
[ 9 of 43] Compiling Gauge.Format     ( Gauge/Format.hs, dist/build/Gauge/Format.o, dist/build/Gauge/Format.dyn_o )
[10 of 43] Compiling Numeric.MathFunctions.Comparison ( math-functions/Numeric/MathFunctions/Comparison.hs, dist/build/Numeric/MathFunctions/Comparison.o, dist/build/Numeric/MathFunctions/Comparison.dyn_o )
[11 of 43] Compiling Numeric.MathFunctions.Constants ( math-functions/Numeric/MathFunctions/Constants.hs, dist/build/Numeric/MathFunctions/Constants.o, dist/build/Numeric/MathFunctions/Constants.dyn_o )
[12 of 43] Compiling Numeric.SpecFunctions.Internal ( math-functions/Numeric/SpecFunctions/Internal.hs, dist/build/Numeric/SpecFunctions/Internal.o, dist/build/Numeric/SpecFunctions/Internal.dyn_o )
[13 of 43] Compiling Numeric.SpecFunctions ( math-functions/Numeric/SpecFunctions.hs, dist/build/Numeric/SpecFunctions.o, dist/build/Numeric/SpecFunctions.dyn_o )
[14 of 43] Compiling Numeric.Sum      ( math-functions/Numeric/Sum.hs, dist/build/Numeric/Sum.o, dist/build/Numeric/Sum.dyn_o )
[15 of 43] Compiling Paths_gauge      ( dist/build/autogen/Paths_gauge.hs, dist/build/Paths_gauge.o, dist/build/Paths_gauge.dyn_o )
[16 of 43] Compiling Gauge.Main.Options ( Gauge/Main/Options.hs, dist/build/Gauge/Main/Options.o, dist/build/Gauge/Main/Options.dyn_o )
[17 of 43] Compiling Statistics.Distribution ( statistics/Statistics/Distribution.hs, dist/build/Statistics/Distribution.o, dist/build/Statistics/Distribution.dyn_o )
[18 of 43] Compiling Statistics.Function ( statistics/Statistics/Function.hs, dist/build/Statistics/Function.o, dist/build/Statistics/Function.dyn_o )
[19 of 43] Compiling Statistics.Internal ( statistics/Statistics/Internal.hs, dist/build/Statistics/Internal.o, dist/build/Statistics/Internal.dyn_o )
[20 of 43] Compiling Statistics.Distribution.Normal ( statistics/Statistics/Distribution/Normal.hs, dist/build/Statistics/Distribution/Normal.o, dist/build/Statistics/Distribution/Normal.dyn_o )
[21 of 43] Compiling Statistics.Math.RootFinding ( statistics/Statistics/Math/RootFinding.hs, dist/build/Statistics/Math/RootFinding.o, dist/build/Statistics/Math/RootFinding.dyn_o )
[22 of 43] Compiling Statistics.Matrix.Types ( statistics/Statistics/Matrix/Types.hs, dist/build/Statistics/Matrix/Types.o, dist/build/Statistics/Matrix/Types.dyn_o )
[23 of 43] Compiling Statistics.Matrix.Mutable ( statistics/Statistics/Matrix/Mutable.hs, dist/build/Statistics/Matrix/Mutable.o, dist/build/Statistics/Matrix/Mutable.dyn_o )
[24 of 43] Compiling Statistics.Quantile ( statistics/Statistics/Quantile.hs, dist/build/Statistics/Quantile.o, dist/build/Statistics/Quantile.dyn_o )
[25 of 43] Compiling Statistics.Sample.Histogram ( statistics/Statistics/Sample/Histogram.hs, dist/build/Statistics/Sample/Histogram.o, dist/build/Statistics/Sample/Histogram.dyn_o )
[26 of 43] Compiling Statistics.Sample.Internal ( statistics/Statistics/Sample/Internal.hs, dist/build/Statistics/Sample/Internal.o, dist/build/Statistics/Sample/Internal.dyn_o )
[27 of 43] Compiling Statistics.Sample ( statistics/Statistics/Sample.hs, dist/build/Statistics/Sample.o, dist/build/Statistics/Sample.dyn_o )
[28 of 43] Compiling Statistics.Matrix ( statistics/Statistics/Matrix.hs, dist/build/Statistics/Matrix.o, dist/build/Statistics/Matrix.dyn_o )
[29 of 43] Compiling Statistics.Matrix.Algorithms ( statistics/Statistics/Matrix/Algorithms.hs, dist/build/Statistics/Matrix/Algorithms.o, dist/build/Statistics/Matrix/Algorithms.dyn_o )
[30 of 43] Compiling Statistics.Transform ( statistics/Statistics/Transform.hs, dist/build/Statistics/Transform.o, dist/build/Statistics/Transform.dyn_o )
[31 of 43] Compiling Statistics.Sample.KernelDensity ( statistics/Statistics/Sample/KernelDensity.hs, dist/build/Statistics/Sample/KernelDensity.o, dist/build/Statistics/Sample/KernelDensity.dyn_o )
[32 of 43] Compiling Statistics.Types.Internal ( statistics/Statistics/Types/Internal.hs, dist/build/Statistics/Types/Internal.o, dist/build/Statistics/Types/Internal.dyn_o )
[33 of 43] Compiling Statistics.Types ( statistics/Statistics/Types.hs, dist/build/Statistics/Types.o, dist/build/Statistics/Types.dyn_o )
[34 of 43] Compiling System.Random.MWC ( mwc-random/System/Random/MWC.hs, dist/build/System/Random/MWC.o, dist/build/System/Random/MWC.dyn_o )
[35 of 43] Compiling Statistics.Resampling ( statistics/Statistics/Resampling.hs, dist/build/Statistics/Resampling.o, dist/build/Statistics/Resampling.dyn_o )
[36 of 43] Compiling Statistics.Resampling.Bootstrap ( statistics/Statistics/Resampling/Bootstrap.hs, dist/build/Statistics/Resampling/Bootstrap.o, dist/build/Statistics/Resampling/Bootstrap.dyn_o )
[37 of 43] Compiling Statistics.Regression ( statistics/Statistics/Regression.hs, dist/build/Statistics/Regression.o, dist/build/Statistics/Regression.dyn_o )
[38 of 43] Compiling Gauge.Monad      ( Gauge/Monad.hs, dist/build/Gauge/Monad.o, dist/build/Gauge/Monad.dyn_o )
[39 of 43] Compiling Gauge.IO.Printf  ( Gauge/IO/Printf.hs, dist/build/Gauge/IO/Printf.o, dist/build/Gauge/IO/Printf.dyn_o )
[40 of 43] Compiling Gauge.Benchmark  ( Gauge/Benchmark.hs, dist/build/Gauge/Benchmark.o, dist/build/Gauge/Benchmark.dyn_o )
[41 of 43] Compiling Gauge.Analysis   ( Gauge/Analysis.hs, dist/build/Gauge/Analysis.o, dist/build/Gauge/Analysis.dyn_o )
[42 of 43] Compiling Gauge.Main       ( Gauge/Main.hs, dist/build/Gauge/Main.o, dist/build/Gauge/Main.dyn_o )
[43 of 43] Compiling Gauge            ( Gauge.hs, dist/build/Gauge.o, dist/build/Gauge.dyn_o )

cbits/cycles.c:55:2: error:
     error: Unsupported OS/architecture/compiler!
   |
55 | #error Unsupported OS/architecture/compiler!
   |  ^
#error Unsupported OS/architecture/compiler!
 ^
1 error generated.
`gcc' failed in phase `C Compiler'. (Exit code: 1)
cabal: Failed to build gauge-0.2.5. See the build log above for details.
```

</details>

--- a/cbits/cycles.c
+++ b/cbits/cycles.c
@@ -1,6 +1,15 @@
 #include "Rts.h"
 
-#if x86_64_HOST_ARCH || i386_HOST_ARCH
+#if darwin_HOST_OS
+
+#include <mach/mach_time.h>
+
+StgWord64 gauge_rdtsc(void)
+{
+  return mach_absolute_time();
+}
+
+#elif x86_64_HOST_ARCH || i386_HOST_ARCH
 
 StgWord64 gauge_rdtsc(void)
 {

--- a/cbits/time-osx.c
+++ b/cbits/time-osx.c
@@ -1,28 +1,10 @@
 #include <mach/mach.h>
-#include <mach/mach_time.h>
 
 #include "gauge-time.h"
 #include "cycles.h"
+#include <time.h>
 
-static mach_timebase_info_data_t timebase_info;
-static double timebase_recip;
-
-void gauge_inittime(void)
-{
-    if (timebase_recip == 0) {
-	mach_timebase_info(&timebase_info);
-	timebase_recip = (timebase_info.denom / timebase_info.numer) / 1e9;
-    }
-}
-
-static inline uint64_t scale64(uint64_t i, uint64_t numer, uint64_t denom)
-{
-    uint64_t high = (i >> 32) * numer;
-    uint64_t low = (i & 0xffffffffULL) * numer / denom;
-    uint64_t highRem = ((high % denom) << 32) / denom;
-    high /= denom;
-    return (high << 32) + highRem + low;
-}
+void gauge_inittime(void) {}
 
 void gauge_record(struct gauge_time *tr)
 {
@@ -32,7 +14,7 @@ void gauge_record(struct gauge_time *tr)
 				                 (task_info_t) &thread_info_data,
 				                 &thread_info_count);
 
-    tr->clock_nanosecs = scale64(mach_absolute_time(), timebase_info.numer, timebase_info.denom);
+    tr->clock_nanosecs = clock_gettime_nsec_np(CLOCK_UPTIME_RAW);
 
     tr->cpu_nanosecs = (((uint64_t) thread_info_data.user_time.seconds) * ref_second) +
                        (((uint64_t) thread_info_data.user_time.microseconds) * ref_microsecond) +
@@ -43,7 +25,7 @@ void gauge_record(struct gauge_time *tr)
 
 double gauge_gettime(void)
 {
-    return mach_absolute_time() * timebase_recip;
+    return clock_gettime_nsec_np(CLOCK_UPTIME_RAW) / 1e9;
 }
 
 static double to_double(time_value_t time)

--- a/changelog.md
+++ b/changelog.md
@@ -1,3 +1,28 @@
+# next
+
+* Change `gauge_rdtsc` to return `mach_absolute_time` on macOS. This is a
+  portable way of returning the number of CPU cycles that works on both Intel-
+  and ARM-based Macs.
+
+* Change `gauge_gettime` to use `clock_gettime_nsec_np` instead of
+  `mach_absolute_time` on macOS. While `mach_absolute_time` has nanosecond
+  resolution on Intel-based Macs, this is not the case on ARM-based Macs, so
+  the previous `mach_absolute_time`-based implementation would return incorrect
+  timing results on Apple silicon.
+
+  There are two minor consequences of this change:
+
+  * Gauge now only supports macOS 10.02 or later, as that is
+    the first version to have `clock_gettime_nsec_np`. As macOS 10.02 was
+    released in 2002, this is unlikely to affect users, but please speak up if
+    this is a problem for you.
+
+  * As `clock_gettime_nsec_np` does not require any special initialization
+    code, `gauge_inittime` is now a no-op on macOS. If you manually invoke
+    the `getTime` function in your code, however, it is still important that
+    you `initializeTime` beforehand, as this is still required for the Windows
+    implementation to work correctly.
+
 # 0.2.5
 
 * Add GHCJS support (statistical analysis is not supported)

